File: tests/fixtures/project_go/utils/crypto.go
package utils

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// HashPassword hashes a password with the given salt using SHA-256.
func HashPassword(password, salt string) (string, error) {
	if password == "" {
		return "", fmt.Errorf("password cannot be empty")
	}
	h := sha256.New()
	h.Write([]byte(password + salt))
	return hex.EncodeToString(h.Sum(nil)), nil
}

// ComparePassword compares a plaintext password against a hashed password.
func ComparePassword(password, salt, hashed string) (bool, error) {
	computed, err := HashPassword(password, salt)
	if err != nil {
		return false, err
	}
	return computed == hashed, nil
}

// GenerateSalt generates a random salt string of the specified byte length.
func GenerateSalt(length int) (string, error) {
	bytes := make([]byte, length)
	_, err := rand.Read(bytes)
	if err != nil {
		return "", fmt.Errorf("failed to generate salt: %w", err)
	}
	return hex.EncodeToString(bytes), nil
}

File: tests/fixtures/project_py/src/utils/crypto.py
from __future__ import annotations

import hashlib
import os


def generate_salt(length: int = 16) -> bytes:
    """Generate a cryptographically secure random salt."""
    return os.urandom(length)


def hash_password(password: str, salt: bytes | None = None) -> tuple[str, bytes]:
    """Hash a password with the given salt, or generate a new salt.

    Returns a tuple of (hashed_password, salt).
    """
    if salt is None:
        salt = generate_salt()
    hashed = hashlib.pbkdf2_hmac("sha256", password.encode("utf-8"), salt, 100_000)
    return hashed.hex(), salt


def compare_password(password: str, hashed: str, salt: bytes) -> bool:
    """Compare a plaintext password against a hashed password."""
    candidate, _ = hash_password(password, salt)
    return candidate == hashed
